sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java

```diff
@@ -0,0 +1,267 @@
+import org.apache.beam.sdk.transforms.Combine.CombineFn;
+import org.apache.beam.sdk.transforms.Sum;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeMatcher;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Tests for {@link CombineFnTester}.
+ */
+@RunWith(JUnit4.class)
+public class CombineFnTesterTest {
+  @Test
+  public void checksMergeWithEmptyAccumulators() {
+    final AtomicBoolean sawEmpty = new AtomicBoolean(false);
+    CombineFn<Integer, int[], Integer> combineFn =
+        new CombineFn<Integer, int[], Integer>() {
+          @Override
+          public int[] createAccumulator() {
+            return new int[] {0};
+          }
+
+          @Override
+          public int[] addInput(int[] accumulator, Integer input) {
+            accumulator[0] += input;
+            return accumulator;
+          }
+
+          @Override
+          public int[] mergeAccumulators(Iterable<int[]> accumulators) {
+            int result = 0;
+            for (int[] accum : accumulators) {
+              if (accum[0] == 0) {
+                sawEmpty.set(true);
+              }
+              result += accum[0];
+            }
+            return new int[] {result};
+          }
+
+          @Override
+          public Integer extractOutput(int[] accumulator) {
+            return accumulator[0];
+          }
+        };
+
+    CombineFnTester.testCombineFn(combineFn, Arrays.asList(1, 2, 3, 4, 5), 15);
+    assertThat(sawEmpty.get(), is(true));
+  }
+
+  @Test
+  public void checksWithSingleShard() {
+    final AtomicBoolean sawSingleShard = new AtomicBoolean();
+    CombineFn<Integer, int[], Integer> combineFn =
+        new CombineFn<Integer, int[], Integer>() {
+          int accumCount = 0;
+
+          @Override
+          public int[] createAccumulator() {
+            accumCount++;
+            return new int[] {0};
+          }
+
+          @Override
+          public int[] addInput(int[] accumulator, Integer input) {
+            accumulator[0] += input;
+            return accumulator;
+          }
+
+          @Override
+          public int[] mergeAccumulators(Iterable<int[]> accumulators) {
+            int result = 0;
+            for (int[] accum : accumulators) {
+              result += accum[0];
+            }
+            return new int[] {result};
+          }
+
+          @Override
+          public Integer extractOutput(int[] accumulator) {
+            if (accumCount == 1) {
+              sawSingleShard.set(true);
+            }
+            accumCount = 0;
+            return accumulator[0];
+          }
+        };
+
+    CombineFnTester.testCombineFn(combineFn, Arrays.asList(1, 2, 3, 4, 5), 15);
+    assertThat(sawSingleShard.get(), is(true));
+  }
+
+  @Test
+  public void checksWithShards() {
+    final AtomicBoolean sawManyShards = new AtomicBoolean();
+    CombineFn<Integer, int[], Integer> combineFn =
```

Comment:
Might be cleaner to just use an Integer for the accumulator as performance doesn't matter here.